### PR TITLE
APP-1416- Camera stream manger

### PR DESCRIPTION
**Pull request by Kschappacher**
Camera stream manager is meant to pull the stream managing logic outside of camera and into separate files. Handling things like video elements stream open and closes as well as handling restarts on reconnect.

To try out this PR use the temp environment found here: https://github.com/viamrobotics/app/pull/1319

### CODE REVIEW

File: web/frontend/viamrobotics-remote-control-0.2.6.tgz
Comment (Kschappacher) [inline web/frontend/viamrobotics-remote-control-0.2.6.tgz:1]: Remove before Merge

File: web/frontend/src/components/camera/camera.vue [hunk 1]
```diff
@@ -1,120 +1,60 @@
...
   client: Client;
   showExportScreenshot: boolean;
   refreshRate: string | undefined;
   triggerRefresh: boolean;
-  toggle:boolean | undefined | null;
+  streamManager:StreamManager;
 }>();
 
-let videoStream = $ref<MediaStream>();
 const imgEl = $ref<HTMLImageElement>();
```
[the remark this answers is not in the document]
Comment (DTCurrie) [reply web/frontend/src/components/camera/camera.vue:25]: That should be follow up work, not for this PR.

**Review by michaellee1019 — CHANGES_REQUESTED** [review]

File: web/frontend/src/components/camera/camera-manager.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+  removeStream () {
+    this.streamCount -= 1;
+    if (this.streamCount === 0) {
+      this.close();
+    }
+  }
+
+  open () {
+    console.log('opening stream');
```
[the remark this answers is not in the document]
Comment (micheal-parks) [reply web/frontend/src/components/camera/camera-manager.ts:38]: We should strip all console.* from prod

File: web/frontend/src/components/camera/camera-manager.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      // Ignore event if received for the wrong stream, in the case of multiple cameras
+      if (eventStream.id !== this.cameraName) {
+        return;
+      }
+      this.VideoElement = eventStream;
+    });
+  }
+
+  close () {
+    console.log('Closing Stream');
```
Comment (micheal-parks) [inline web/frontend/src/components/camera/camera-manager.ts:54]: ✂️

File: web/frontend/src/components/camera/stream-manager.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import type {
+  Client,
+} from '@viamrobotics/sdk';
+import { CameraManager } from './camera-manager';
+
+export class StreamManager {
+  isConnected: boolean;
```
Comment (micheal-parks) [inline web/frontend/src/components/camera/stream-manager.ts:7]: What is the purpose of this if it's always true?

File: web/frontend/cypress/e2e/robot-components/base.cy.ts [hunk 5]
```diff
@@ -11,20 +11,19 @@ describe('base', () => {
 
     // Select camera
     cy.get('[aria-label="Refresh frequency for test_camera"]').click();
-    cy.get('[data-parent="base"').find('video');
```
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/cypress/e2e/robot-components/base.cy.ts:14]: I couldn't actually get cy.get('[aria-label="Refresh frequency for test_camera"]').click(); to properly click on the button. so the video element never shows up. We got this to "Work" in the patch pr by keeping the camera element as a v-show

[the remark this answers is not in the document]
Comment (michaellee1019) [reply web/frontend/cypress/e2e/robot-components/base.cy.ts:14]: Yeah that selector was giving me pain too. Maybe we should add something (class name, id, etc) to make it easily selectable. I'd also be curious to try and select this with the Playwright recording tool and see what selector is recommended. CC @anjinai

File: web/frontend/src/components/camera/camera-manager.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: removed all the console logs i added

File: web/frontend/src/components/camera/camera-manager.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+  stream:StreamClient;
+
+  public VideoElement:MediaStream;
+
+  constructor (cameraName:string, client:Client) {
+    this.cameraName = cameraName;
+    this.client = client;
+    this.streamCount = 0;
+    this.stream = new StreamClient(client);
```
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: Totes agree, I changed things so stream manager creates the streamClient and passes it to the camera managers

File: web/frontend/src/components/camera/camera-manager.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+export class CameraManager {
+  cameraName:string;
+
+  client:Client;
+
+  streamCount:number;
+
+  stream:StreamClient;
+
+  public VideoElement:MediaStream;
```
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: I changed it slightly differently
stream->streamClient
VideoElement->videoStream

File: web/frontend/src/components/camera/camera-manager.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      this.VideoElement = eventStream;
+    });
+  }
+
+  close () {
+    console.log('Closing Stream');
+    this.stream.remove(this.cameraName);
+  }
+
+  async setImageElement (imgEl:HTMLImageElement|undefined) {
```
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: Done

File: web/frontend/src/components/camera/camera-manager.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    let blob;
+    try {
+      blob = await new CameraClient(this.client, this.cameraName).renderFrame('image/jpeg');
```
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: Yupp created client on manager construction instead. I could also define it the first time setImageSrc is called instead if that makes more sense

File: web/frontend/src/components/camera/stream-manager.ts (hunk 4 above)
Comment (Kschappacher) [reply web/frontend/src/components/camera/stream-manager.ts]: Opps logic from a forgotten idea. It has been scrubbed

File: web/frontend/src/components/camera/stream-manager.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+export class StreamManager {
+
+  cameraManagers: Map<string, CameraManager>;
+
+  client: Client;
+
+  streamClient:StreamClient;
+
+  constructor (client:Client) {
```
Comment (DTCurrie) [inline web/frontend/src/components/camera/stream-manager.ts:15]: Similar nit here:

```typescript
  private streamClient:StreamClient;

  constructor (
    private client:Client,
    private cameraManagers: Map<string, CameraManager> = new Map<string, CameraManager>()

  ) {
    this.streamClient = new StreamClient(client);
  }
  ```

File: web/frontend/src/components/camera/camera-manager.ts [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+  client:Client;
+
+  streamCount:number;
+
+  streamClient:StreamClient;
+
+  public videoStream:MediaStream;
+
+  constructor (cameraName:string, client:Client, streamClient:StreamClient) {
```
[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: That looks way cleaner, just made the changes!

[the remark this answers is not in the document]
Comment (Kschappacher) [reply web/frontend/src/components/camera/camera-manager.ts]: That looks way cleaner, just made the changes!

File: web/frontend/src/components/camera/stream-manager.ts (hunk 10 above)
Comment (Kschappacher) [reply web/frontend/src/components/camera/stream-manager.ts]: and here too

File: web/frontend/src/components/camera/stream-manager.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+  constructor (
+    private client:Client,
+    public cameraManagers: Map<string, CameraManager> = new Map<string, CameraManager>()
+
+  ) {
+    this.streamClient = new StreamClient(client);
+  }
+
+  setCameraManager (cameraName:string) {
+    const tempManager = new CameraManager(cameraName, this.client, this.streamClient);
```
Comment (DTCurrie) [inline web/frontend/src/components/camera/stream-manager.ts:19]: To go with the above changes.

```suggestion
    const tempManager = new CameraManager(this.client, cameraName, this.streamClient);
```

**Review by DTCurrie — APPROVED** [review]
✅ even more nits but very trivial

**Review by michaellee1019 — APPROVED** [review]
LGTM!

**Pull request merged**